### PR TITLE
Database interactions for relayer

### PR DESCRIPTION
**Pull request by sczembor**
## Functions to insert, update and query from database + unit tests

closes: #30 

---

### Author Checklist

_All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues._

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] added appropriate labels to the PR
- [ ] targeted the correct branch (see [PR Targeting](https://github.com/umee-network/umee/blob/main/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary

### CODE REVIEW

File: native/database/db.go [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+	}
+
+	_, err = db.Exec(`
+        CREATE TABLE IF NOT EXISTS transactions (
+            txid TEXT PRIMARY KEY,
+            rawtx TEXT NOT NULL,
+            status TEXT NOT NULL
+        )
+    `)
+	if err != nil {
```
Comment (vuvoth) [inline native/database/db.go:45]: Why we don't return error directly here?

File: native/database/db.go [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+		return fmt.Errorf("failed to create table: %w", err)
+	}
+	return nil
+}
+
+// InsertTransaction inserts a new transaction into the database
+func InsertTransaction(tx Transaction) error {
+	stmt, err := db.Prepare("INSERT INTO transactions(txid, rawtx, status) values(?,?,?)")
```
Comment (vuvoth) [inline native/database/db.go:53]: Can we make the SQL string constant?

File: native/database/db.go (hunk 1 above)
Comment (sczembor) [reply native/database/db.go:45]: Maybe I should use the logger there to log the error and just return it? What do you think?

Comment (vuvoth) [reply native/database/db.go:45]: I think return error more better in this case.

Comment (sczembor) [reply native/database/db.go:45]: I wrapped the error to provide more details. But if you think just returning the error is better I will fix it

Comment (sczembor) [reply native/database/db.go:53]: Done

File: native/database/db.go (hunk 2 above)
Comment (sczembor) [reply native/database/db.go]: Done

File: native/database/db.go [hunk 3]
```diff
@@ -0,0 +1,132 @@
+package database
```
Comment (robert-zaremba) [inline native/database/db.go:1]: native should be a package to connect to Native chain. The relayer DB should be in a separate package, so let's move it to the root.

File: native/database/db.go [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+)
+
+// TransactionStatus represents the different states of a transaction.
+type TransactionStatus string
+
+// Different tx states
+const (
+	StatusPending     TransactionStatus = "pending"
+	StatusBroadcasted TransactionStatus = "broadcasted"
+	StatusConfirmed   TransactionStatus = "confirmed"
```
Comment (robert-zaremba) [inline native/database/db.go:16]: let's use numeric values for enum.

File: native/database/db.go [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+
+	_, err = db.Exec(createTransactionsTableSQL)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// InsertTransaction inserts a new transaction into the database
+func InsertTransaction(tx Transaction) error {
```
Comment (robert-zaremba) [inline native/database/db.go:60]: let's use shorter names: InsertTx, UpdateTx ...

File: native/database/db_test.go [hunk 6]
```diff
@@ -0,0 +1,84 @@
+package database
+
+import (
+	"testing"
+)
+
+func TestInsertTransaction(t *testing.T) {
+	err := InitDB(":memory:") // in-memory database
+	if err != nil {
```
Comment (robert-zaremba) [inline native/database/db_test.go:9]: use https://pkg.go.dev/gotest.tools/assert

File: native/database/db_test.go [hunk 7]
```diff
@@ -0,0 +1,84 @@
+package database
+
+import (
+	"testing"
+)
+
+func TestInsertTransaction(t *testing.T) {
+	err := InitDB(":memory:") // in-memory database
```
Comment (robert-zaremba) [inline native/database/db_test.go:8]: let's define a helper method: `initTestDB()`

File: native/database/db_test.go [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+	}
+
+	pendingTxs, err := GetPendingTransactions()
+	if err != nil {
+		t.Errorf("GetPendingTransactions() error = %v", err)
+	}
+
+	if len(pendingTxs) != 2 {
+		t.Errorf("Expected 2 pending transactions, got %d", len(pendingTxs))
+	}
```
Comment (robert-zaremba) [inline native/database/db_test.go:50]:
```suggestion
	assert.Equal(t, len(pendingTxs), 2, "expected 2 pending transactions")
```

File: native/database/db.go [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+const (
+	insertTransactionSQL       = "INSERT INTO transactions(txid, rawtx, status) values(?,?,?)"
+	getPendingTransactionsSQL  = "SELECT txid, rawtx, status FROM transactions WHERE status = ?"
+	getTransactionByTxidSQL    = "SELECT txid, rawtx, status FROM transactions WHERE txid = ?"
+	updateTransactionStatusSQL = "UPDATE transactions SET status = ? WHERE txid = ?"
+	createTransactionsTableSQL = `
+        CREATE TABLE IF NOT EXISTS transactions (
+            txid TEXT PRIMARY KEY,
+            rawtx TEXT NOT NULL,
+            status TEXT NOT NULL
```
[the remark this answers is not in the document]
Comment (sczembor) [reply native/database/db.go:29]: Done

File: native/database/db.go [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+
+// Transaction represents a transaction record in the database.
+type Transaction struct {
+	BtcTxID uint64            `json:"txid"`
+	RawTx   string            `json:"rawtx"`
+	Status  TransactionStatus `json:"status"`
+	// TODO: other fields
+}
+
+var db *sql.DB
```
[the remark this answers is not in the document]
Comment (sczembor) [reply native/database/db.go:42]: Done

File: native/database/db.go (hunk 4 above)
Comment (sczembor) [reply native/database/db.go:16]: Done

File: native/database/db.go (hunk 5 above)
Comment (sczembor) [reply native/database/db.go:60]: Done

File: native/database/db_test.go (hunk 6 above)
Comment (sczembor) [reply native/database/db_test.go:9]: Done

File: native/database/db_test.go (hunk 7 above)
Comment (sczembor) [reply native/database/db_test.go:8]: Done

File: database/db.go [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// GetTx retrives a transaction by its txid
+func (db DB) GetTx(txID uint64) (*Tx, error) {
+	stmt, err := db.conn.Prepare(getTransactionByTxidSQL)
```
Comment (robert-zaremba) [inline database/db.go:84]: You don't have to use `Prepare` if you don't reuse the `stmt`. Either you store the statement in `db` or you use `db.QueryRow` directly.

File: database/db.go [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+	var err error
+	db.conn, err = sql.Open("sqlite3", dbPath)
+	if err != nil {
+		return nil, err
+	}
+
+	_, err = db.conn.Exec(createTransactionsTableSQL)
+	if err != nil {
+		return nil, err
+	}
```
Comment (robert-zaremba) [inline database/db.go:61]: Let's move DB init outside of DB creation.

File: database/db.go [hunk 13]
```diff
@@ -0,0 +1,139 @@
...
+}
+
+// DB holds the database connection and provides methods for interacting with it.
+type DB struct {
+	conn *sql.DB
+}
+
+// NewDB creates a new DB instance and initializes the database connection.
+func NewDB(dbPath string) (*DB, error) {
+	db := &DB{} // Create a new DB instance
```
Comment (robert-zaremba) [inline database/db.go:49]:
```suggestion
	db := &DB{}
```

Comment (robert-zaremba) [reply database/db.go:49]: avoid "comment-code redundancy"

File: database/db.go [hunk 14]
```diff
@@ -0,0 +1,139 @@
...
+
+	// Initialize the database connection
```
Comment (robert-zaremba) [inline database/db.go:51]:
```suggestion
```

File: database/db.go [hunk 15]
```diff
@@ -0,0 +1,139 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	_, err = db.conn.Exec(createTransactionsTableSQL)
+	if err != nil {
+		return nil, err
+	}
+
+	return db, nil // Return the initialized DB instance
```
Comment (robert-zaremba) [inline database/db.go:63]:
```suggestion
	_, err = db.conn.Exec(createTransactionsTableSQL)
	return db, err
```

File: database/db.go [hunk 16]
```diff
@@ -0,0 +1,139 @@
+package database
+
+import (
+	"database/sql"
+
+	_ "github.com/mattn/go-sqlite3" // Import the SQLite driver
```
Comment (robert-zaremba) [inline database/db.go:6]:
```suggestion
	_ "github.com/mattn/go-sqlite3"
```

Comment (robert-zaremba) [reply database/db.go:6]: avoid "comment-code redundancy"

**Review by robert-zaremba — COMMENTED** [review]
pre-approving
- please rename `database` package to `dal` as suggested in the previous review
- avoid "comment-code redundancy"

**Review by robert-zaremba — APPROVED** [review]
pre-approving, see the previous review

File: database/db.go [hunk 17]
```diff
@@ -0,0 +1,139 @@
+package database
```
[the remark this answers is not in the document]
Comment (sczembor) [reply database/db.go:1]: Done

File: database/db.go (hunk 16 above)
Comment (sczembor) [reply database/db.go:6]: I removed the other unnecessary comments, however this is required by the linter (blank import)

File: database/db.go (hunk 12 above)
Comment (sczembor) [reply database/db.go]: Done

File: database/db.go (hunk 11 above)
Comment (sczembor) [reply database/db.go]: Done

**Pull request merged**